### ISSUE TITLE
[v6] Discussion: Impact on SSR after removal of <Redirect />

### ISSUE THREAD

**Issue by pseudo-su** [ReactTraining/react-router#7267]
I'm trying to reconcile how I've been using the `<Redirect />` component and figure out how the removal of `<Redirect />` is meant to be translated into my "universal" react applications.

tl;dr the removal of `<Redirect />` will break the mechanism I've been using to have a consistent redirect experience across server-rendering and client-rendering contexts. I'd like to put my case forward as to why the justification for removing `<Redirect />` baffles me. Particularly the suggestion that "you should be doing it on the server anyway", because server-side redirects are precisely one of my use-cases for `<Redirect />`.

The way I have been using `<Redirect />` in the past on the server looks something like this (This won't work as-is but I'm just trying to articulate the gist of the important bits)

```js
// universal/app.js
function App() {
  return (
    <Switch>
      <Route exact path="/" render={() => <Redirect to="/thing1" />} />
      <Route path="/thing1" component={Thing1} />
      <Route path="/thing2" component={Thing2} />
      <Route component={Error404} />
    </Switch>
  )
}

// client/index.js
const mountElement = document.getElementById('root')
hydrate(
  <BrowserRouter>
    <App />
  </BrowserRouter>
  mountElement
)

// server/index.js
function handler(request, response) {
  const reactRouterContext = {};
  const renderedPage = render(
    <StaticRouter location={request.originalUrl} context={reactRouterContext}>
      <App />
    </StaticRouter>
  )

  // Check if the router context contains a redirect. if it does, we need to set
  // the specific status and redirect header and end the response.
  if (reactRouterContext.url) {
    response.status = 302;
    response.set('Location', reactRouterContext.url);
    return;
  }

  response.status = reactRouterContext.missed ? 404 : 200;
  response.body = wrapPageInHTMLTemplate(renderedPage);
}

```

The recommendation from change log of the v6 alpha is:
> If you really need to redirect on the initial render, you can either a) do it on your server (probably best, so it can be cached at the HTTP level instead of doing it in every user's browser) or b) do it outside of React Router (e.g. using the history API directly).

It sounds like it was removed with the assumption it would only affect people using it for client-side redirects... but I'm specifically using `<Redirect />` as a mechanism to have a unified redirect experience regardless of whether it's the first server-side render or subsequent client-side renders.

The wording around needing to "do it on your server" implies to me that server-rendering is viewed to be outside the scope of `react-router`. Is this signposting a deliberate move away from having SSR capability in `react-router` going forward?

Ultimately if that's the direction this project is going in I can work around it, but I'm looking for clarification first.

EDIT:
I found where this is currently mentioned in the docs so I'll add it here for completeness sake https://reacttraining.com/react-router/web/guides/server-rendering

**Comment by aappddeevv** [comment]
Why would you need to change that (other than syntax and how the components used)? Instead of Redirect there is a Navigate component.

**Comment by pseudo-su** [comment]
Are you asking why I can't use `<Navigate />` in the same way that I describe using my current usage of `<Redirect />`?

Apart from it being semantically incorrect there's a comment/warning here stating

> `<Navigate>` must not be used on the initial render in a <StaticRouter>. This is a no-op, but you should modify your code so the <Navigate> is only ever rendered in response to some user interaction or state change.

https://github.com/ReactTraining/react-router/blob/v6.0.0-alpha.3/packages/react-router/index.js#L108

**Comment by aappddeevv** [comment]
Ah, got it.

**Comment by pseudo-su** [comment]
As of the latest alpha:

> Removed the <StaticRouter context> API. We don't support navigation on the initial render in v6, so this API is unnecessary.

I find this confusing because I don't understand the static router context as being related to navigating during the initial render.

EG: application renders a 404 page in a StaticRouter and the SSR wrapper needs to know that it should set the statusCode on the response.

**Comment by max-mykhailenko** [comment]
I have the same question and one more about SSR. Previously I was able to get matched route with `matchPath`, but now as I understand you changed path ideology and I should use another methods. Need your advice

**Comment by dimaqq** [comment]
>  justification for removing ... baffles me

For the less involved, can you link to the original idea to remove?

**Comment by pseudo-su** [comment]
@dimaqq The only information I'm going by is the alpha release notes

- https://github.com/ReactTraining/react-router/releases/tag/v6.0.0-alpha.3
- https://github.com/ReactTraining/react-router/releases/tag/v6.0.0-alpha.4

**Comment by pseudo-su** [comment]
Just to be clear @Askadias, I believe that `StaticRouter` hasn't been removed entirely, but the ability to pass it `context` to it has. Which is a problem because it's the [current canonical way](https://reacttraining.com/react-router/web/guides/server-rendering/404-401-or-any-other-status) to facilitate rendering different HTTP Status codes with `react-router` in a universal react app.

FYI: I was somewhat hoping that there might be some clarification with the [remix run newsletter](https://twitter.com/ryanflorence/status/1270845718865965057?s=20):

> [...etc]
> 
> All hosting services have a way to define redirects for URLs that have changed. Most React apps just set up some client side redirects with React Router and call it a day. The problem is the search engine bot gets a 200, when it should get a 30x! Now your page rank is going to get effected negatively for those URLs that have moved. For some projects, this might mean a huge hit to revenue, signups, etc.
>
> Instead of configuring this client side, you should configure it on your server. That way the user and the search engine bots get a 30x at the old URL and then a 200 at the new one.
>
>However, if you only configure you redirect on the server, then any links in your app that haven't been updated will not get redirected when the user clicks them even though it's a valid page.
>
>Holy smokes ... so now, to do this properly, you have to configure redirects on the server AND the client.
>
>But ... that's still not enough 😅 If a bot follows a redirected link, the page will update client side but the bot will not get the proper status codes!
>
>Perhaps this is why few people tackle it, it's a little wild.
>
>In Remix, you only have to configure your redirects on your server, using whatever API your platform or host provides. Usually there's a JSON file to configure them. If the user lands on a redirected page, the server is the only thing involved and does the right thing. If the user or a bot clicks a link to a redirect page, just like in scenario (3), Remix doesn't match any routes in the client, so it goes back to the server with a new page load, then the server takes over and sends the right status code.

I'm still primarily confused about why there's an assumption that it needs to be "configured once on the client and once on the server". My use case for `<Redirect />` is precisely so that I only have to configure it once but get *both* server and client redirect behavior.

**Comment by Askadias** [comment]
My use-case is pretty much similar. We apply a lot of various async logic based on query string in our OAuth application using redux. As result we get a redirect url and render is in router using `<Redirect to={redirectUrl}/>`. If we apply the same async precondition on server we get the final redux state and as a result retrieve the redirection url from the router context.

**Comment by suprMax** [comment]
What is the latest status of this? Does someone know if RR6 will support SSR?

**Comment by capaj** [comment]
@suprMax I don't know, but removing SSR support from RR6 would be a major issue for a lot of people. I don't think anybody wants that.

**Comment by suprMax** [comment]
I'm cautiously optimistic that it won't be removed, I haven't tried RR6 myself yet. I think the amount of effort we've put at our company in keeping up with RR updates pretty much equals what would take to roll our own solution, so this update we will probably drop RR altogether if it's anything like 3->4. I wanted a bit of a heads up what to plan for.

**Comment by Zertz** [comment]
From the example in #7535, I ended up with something similar to this:

```jsx
<Routes>
  <Route path="a" element={<A /> />
  <Route path="b" element={<B /> />
  <Route element={<Navigate to="a" />} />
</Routes>
```

It seems to work correctly and is almost equivalent to `<Redirect />`

**Comment by pseudo-su** [comment]
@Zertz `<Navigate />` is a perfect adequate replacement for `<Redirect />` when rendering on the client.

The problem is that `<Naviage />` is a no-op on the server and therefore doesn't cater to the SSR use cases that the current `<Redirect />` component does.

[Docs here](https://reacttraining.com/react-router/web/guides/server-rendering) here for more details of how `<Redirect />` currently works on the server

**Comment by rstrand** [comment]
@pseudo-su would it be an option to create your own `<Redirect />` component which calls `<Navigate />` on the client side and implements your own redirect status code rendering on the server?

**Comment by pseudo-su** [comment]
> 
> 
> @pseudo-su would it be an option to create your own `<Redirect />` component which calls `<Navigate />` on the client side and implements your own redirect status code rendering on the server?

Yes, it's possible. The reason I opened this issue is because I wasn't sure if this side-effect is/was a deliberate deprecation or not. Personally I think if the plan is to deprecate/remove the [current documented way of doing SSR with react router](https://reactrouter.com/web/guides/server-rendering) it'd be a really good idea to at least update the current docs to clarify that things to do with SSR are going to be removed from `react-router` going forward.

That was the point of my original question:

> Is this signposting a deliberate move away from having SSR capability in react-router going forward?

For what it's worth, I'm assuming the silence means that the change **will** be made and server rendering will be removed from `react-router`.

It's not a *massive* inconvenience, but the lack of clarity makes it confusing as to if people need to do anything. In case anyone's particularly curious, this is what I did to prepare for the removal of `<Redirect />` and `staticContext`.

https://github.com/bootleg-rust/sites/tree/9028da0c40d1261a407479c61dd8c5c27ae47453/packages/lib-ssr-toolbox/src/http

tl;dr basically I have a seperate `HttpProvider` and have copied some of the current implementation of <Redirect /> etc:

```tsx
// universal/app.js
function Dashboard() {
  return <>
    Home page
    <CacheControl maxAge={200} />
  </>
}

function Error404() {
  return <>
    <HttpStatus code={400} />
    {/* etc */}
  </>
}

function App() {
  return (
    <Switch>
      <Route exact path="/" render={() => <Redirect to="/dashboard" />} />
      <Route path="/dashboard" component={Dashboard} />
      <Route path="/thing2" component={Thing2} />
      <Route component={Error404} />
    </Switch>
  )
}

// client/index.js
const mountElement = document.getElementById('root')
hydrate(
  <BrowserRouter>
    <App />
  </BrowserRouter>
  mountElement
)

// server/index.js
function handler(ctx: koa.Context) {
  const httpContext: HttpContextData = {
      cacheControl: [],
      statusCode: 200,
    };
  const renderedPage = render(
    <HttpProvider context={httpContext}>
      <StaticRouter location={ctx.request.url}>
        <App />
      </StaticRouter>
    </HttpProvider>
  )

  // Redirect when <Redirect /> is rendered
  if (httpContext.redirectLocation) {
    // Somewhere a `<Redirect>` was rendered
    ctx.redirect(httpContext.redirectLocation);
    return;
  }

  // Handle status codes
  ctx.status = 200;
  if ([400, 401, 402, 403, 404].includes(httpContext.statusCode)) {
    ctx.status = httpContext.statusCode;
  }
  // Set cache-control based on rendered content
  const cacheControl = reconcileCacheControlOptions(httpContext);
  const cacheControlValue = format(cacheControl);
  if (cacheControlValue) {
    ctx.set("Cache-Control", cacheControlValue);
  }

  ctx.response.body = wrapPageInHTMLTemplate(renderedPage);
}
```

**Comment by suprMax** [comment]
We will implement RR alternative with full SSR support and API similar to RR5. I will link here once we have stable implementation.

**Comment by t-pyrope** [comment]
For me worked to pass "*" as a path of the last route (`path="*"`), without that it didn't work
```
<Routes>
    <Route path="home" element={<HomePage />} />
    <Route path="*" element={<Navigate to="home" />} />
</Routes>
```

**Comment by calvinaco** [comment]
If you have defined the parent `Route` with a layout, you can try the following code snippet
```
<Routes>
    <Route path="/" element={<Layout />}>
        <Route path="home" element={<HomePage />} />
        <Route index element={<Navigate to="home" />} />
    </Routes>
</Routes>
```

**Comment by lantanios** [comment]
I am getting the same error: 
`<Navigate> must not be used on the initial render in a <StaticRouter>. This is a no-op, but you should modify your code so the <Navigate> is only ever rendered in response to some user interaction or state change.`

I am using latest packages of react(17.0.2)/react-router(6.1.1)  with SSR.

To test the flows provided in documentation I combined example/guide from the react-router  https://v5.reactrouter.com/web/example/auth-workflow and https://v5.reactrouter.com/web/guides/server-rendering

and getting the same error **here is the code**: https://stackblitz.com/edit/github-tvbwjh-rsabzt?devtoolsheight=33&file=src/App.tsx

To see the error you will need to navigate to /protected route while not logged in and check for server log: https://github-tvbwjh-rsabzt--3000.local.webcontainer.io/protected

**Comment by t-fritsch** [comment]
Sorry to dig-up this old issue, but it has been automatically closed 3 years ago, and as far as I understand the discussion there is still no built-in way -_without a "Data Router"_- to detect redirection server side as it was possible with the `context` before, shouldn't this issue be reopened and worked on ? Did I miss some solution to the problem ?
